README.md
# Configuration Management CI/CD

**Mục lục:**

[1. Cài đặt bundle cho project](#1-cài-đặt-bundle-cho-project)

[2. Using Xcode Configuration (.xcconfig) to Manage Different Build Settings](#2-using-xcode-configuration-xcconfig-to-manage-different-build-settings)

[3. Basic config Travis CI](#3-basic-config-travis-cI)

[4. Auto Review Code and Report](#4-auto-review-code-and-report)

[5. Certificate và Provisioning profile](#5-certificate-và-provisioning-profile)

[6. Fastlane đại cương](#6-fastlane-đại-cương)

[7. Deploy app](#7-deploy-app)

[8. Trigger build](#8-trigger-build)

​	Ps: Bạn có thể vào thư mục `Documents` để xem từng phần riêng lẻ.

---



### 1. Cài đặt bundle cho project

Cài đặt `bundle` (nếu chưa có), mở Terminal lên và gõ

``` ruby
gem install bundle
```

Nếu chưa cài đặt luôn *gem* thì làm ơn lên Google tìm cách nha

* **Bước 1: init gemfile**

```ruby
bundle init
```

Nó sẽ tạo ra 1 file tên là `gemfile` . 

* **Bước 2: edit gemfile**

Dùng trình editor nào đó mà mở lên edit lại, ví dụ như sau:

```ruby
gem 'fastlane', '2.97.0'
gem 'xcpretty', '0.2.8'
gem 'cocoapods', '1.5.3'
gem 'linterbot', '0.2.7'
gem 'slather', '2.4.5'
```

* **Bước 3: install gem**

Sinh ra file `gemfile.lock` 

```ruby
bundle install
```

* **Bước 4: install cocoapod**

Bạn trước tiên sẽ edit và thêm `cocoapod` vào gemfile, sau đó chạy lệnh sau để tạo cocoapod

```ruby
bundle exec pod init
```

Tạo ra file `Podfile`. Qua lệnh trên bạn sẽ thấy để dùng `bundle` thực thi 1 thứ gì đó thì bạn cần có `exec` , sau đó là nhưng thứ gì bạn muốn thực hiện

* **Bước 4: edit Podfile**

Bạn mở file `Podfile` lên và thêm các thư viện mà bạn cần, như bao lần trước đây khi bạn làm với dự án.

* **Bước 5: install pod**

Tiếp tục sử dụng `bundler exce pod` để thực hiện các lệnh của `pod` thông qua `bundle` 

```ruby
bundle exce pod install
```

Lệnh này tương tự `pod install` nó sẽ tạo ra `Podfile.lock` và file `workspace` cho project của bạn.

#### Dùng đầu óc tưởng tượng thêm xí:

- CocoaPod là chỗ chứa các thư viện cho iOS: muốn sài thì phải:

   `init` --> edit *podfile* --> `pod install`

- Bundle thì là chỗ chứa các thứ trâu chó như là *cocoapod* và nó cũng làm tương tự như thế : 

  `init` --> edit gemfile --> `install` --> thêm 1 bữa nữa là muốn chạy gì thì gọi lệnh `exec` mà chạy

---



### 2. Using Xcode Configuration (.xcconfig) to Manage Different Build Settings

Đọc tham khảo tại đây : <https://www.appcoda.com/xcconfig-guide/>

#### 2.1 Creating the Build Configuration

* Thêm configuration cho project, thường sẽ là:

- - Debug
  - Staging
  - Release

* Việc này đơn giản là Duplicate 1 cái có sẵn trong Project

#### 2.2 Using Xcode Configuration File (.xcconfig)

* Về Xcode Configuration file (. xcconfig) chứa những cặp giá trị (key/value) dùng trong project

* New file (**.xcconfig**) —> tạo 3 file cho 3 môi trường —> set vào các config đã đc tạo ra

* Tiến hành edit hay thêm các cặp giá trị vào các file config. Chú ý tất cả các key phải đều đc có mặt ở tất cả các file config

*Ví dụ:*

​	**Debug:**

```swift
IS_APP_NAME = Project (dev)

IS_APP_VERSION = 0.3

IS_APP_BUNDLE_ID = com.fx.TheLastProject.Debug
```



​	**Staging**

```swift
IS_APP_NAME = Project (Staging)

IS_APP_VERSION = 0.2

IS_APP_BUNDLE_ID = com.fx.TheLastProject.Staging
```

​	**Release:**

```swift
IS_APP_NAME = The Last Project

IS_APP_VERSION = 0.1

IS_APP_BUNDLE_ID = com.fx.TheLastProject.Release
```

Mình có thể change cả icon (cái này để sau đi hì)

#### 2.3 Accessing Variables from Code

* Đây là phần cần quan tâm, vì chứa nhiều thứ quan trọng như: Key, Endpoint …

  **Debug.xcconfig:**

```swift
BACKEND_URL = http:\/\/api.intensifystudio.com/development

CONSUMER_KEY = ck_a57e4fa2e14c12ae3f400371cf2951ec3dea5_dev

CONSUMER_SECRET = cs_c847caa35ce1041e9c69d239141f13f63bb22b_dev
```

​	**Staging.xcconfig:**

```swift
BACKEND_URL = http:\/\/api.intensifystudio.com/staging

CONSUMER_KEY = ck_a57e4fa2e14c12f400371cf2951ec3dea5_staging

CONSUMER_SECRET = cs_c847caa35ce1041e9c69d239141f13f63bb22b_staging
```

​	**Release.xcconfig:**

```swift
BACKEND_URL = http:\/\/api.intensifystudio.com/

CONSUMER_KEY = ck_a57e4fa2e14c12f400371cf2951ec3dea5

CONSUMER_SECRET = cs_c847caa35ce1041e9c69d239141f1f63bb22b
```

XONG —> vào file Info.plist mà thêm chúng vào (add row)

Lấy giá trị từ file *Info.plist* vào trong code như sau:

```swift
func infoForKey(_ key: String) -> String? {
        return (Bundle.main.infoDictionary?[key] as? String)?
            .replacingOccurrences(of: "\\", with: "")
 }
```

#### 2.4  Switching Between Build Configurations

Vào scheme -> Edit scheme -> Run -> Build Config : mà thay đổi

---



### 3. Basic config Travis CI

Phần này dành cho các bạn mù về CI, nếu bạn đã trùm thì có thể bỏ qua phần này. Ahihi. Trước tiên thì đi vào mấy khái niệm cơ bản trước đã:

* **CI là gì?**
  - Viết tắt của ***Continuous Integration (CI)*** ; gọi là tích hợp liên tục
  - Giúp cho mình trong việc kiểm tra, build app, test (UI và Unit), review code giúp
  - Tích hợp vào GitHub để thông báo cho dev biết tình trạng các pull request của mình
  - Về iOS thì CI khá đắt do giá máy MAC OS trên các dịch vụ CI tốn phí nhiều. 
* **Travis CI**
  * Là dịch vụ CI cho lập trình viên
  * Tích hợp đc với Github
  * Có 2 phiên bản
    * Free (chỉ chạy đc public repo) : <https://travis-ci.org/>
    * Trả phí (chạy đc private repo) : <https://travis-ci.com/>

#### 3.1 The YAML File

- Là file có đuôi là `yml` 

- Các dịch vụ CI sẽ tự động clone source code từ repo của bạn về, nếu trong repo có `yaml file` thì sẽ tự động đọc file đó và thực thi các lệnh hay setup cấu hình mà mình ghi trong đó.

- Với Travis CI thì file sẽ tên là `.travis.yml`

  Ví dụ:

  ```ruby
  language: objective-c
  osx_image: xcode10.1
  xcode_workspace: TheLastProject.xcworkspace
  xcode_scheme: TheLastProject_Development
  xcode_destination: platform=iOS Simulator,OS=12.1,name=iPhone X
  before_install:
    - bundle install
    - bundle exec pod install --repo-update
  ```

#### 3.2 Cấu trúc cơ bản của YAML File

* **language**

  * Travis hỗ trợ cả 2 ngôn ngữ `Objective-C` và `Swift`
  * Tuy nhiên cài đặt thì `objective-c` cho cả 2

* **osx_image**

  * Đây là máy ảo chọn để build project lên mình
  * Có nhiều cấu hình máy áo, bạn nên chọn loại phù hợp với project của mình. Nếu như chọn máy ảo đời thấp hơn so với yêu cầu project thì sẽ ko build đc
  * Danh sách các máy ảo đc support của Tracis CI sau [đây](https://docs.travis-ci.com/user/reference/osx/)

* **xcode_workspace**

  * Vì có nhiều project trong 1 workspace (Pod, project hiện tại, các thư viện nếu build trực tiếp từ project)
  * Nên hãy chọn cho đúng
  * Nếu có thư mục thì hãy trỏ đường dẫn vào. Ví dụ:

  ```ruby
  ./Example/TravisCIBlog.xcworkspace
  ```

* **xcode_scheme**

  * Cực kì quan trọng vì nó sẽ quyết định bạn muốn build gì hay kiểm tra gì. Hãy chọn cho đúng.
  * Ngoài ra, thì cần phải ***share*** scheme để CI có thể build đc (vào Scheme -> Managment —> xem vào chọn)

* **xcode_destination**

  * Lựa chọn cho các thanh niên nhà nghèo khi dùng `iPhone simulator`
  * Bạn có thể quyết định chọn máy ảo nào để build
  * Danh sách các máy ảo được support ở [đây](https://gist.github.com/jgsamudio/4a38d468c12aaec84cdc5f5b2c77b726)

* **before_install**

  * Cực kì quan trọng, vì nó sẽ chạy các lệnh cơ bản trước khi thực hiện
  * Vì máy ảo đó trên mạng nên ko có như local là bạn chỉ cần chạy các lệnh như `pod` rồi cứ thế mà build tới. Nên mình phải thêm các lệnh đó vào phần này

  ```ruby
  before_install:
    - bundle install
    - bundle exec pod install --repo-update
  ```

  * Nếu có lệnh nào phức tạp hơn thì hay chuẩn bị gì thêm thì bỏ thêm vào

#### 3.3 Triggering a Build

* Phần này thì mặc định Tracis CI sẽ theo dõi các `pull request` vào `master`
* Nên bạn cứ tạo `new branch` và gởi `pull request` . Các lần sau chỉ cần `commit` và `push` vào `branch` đó thì CI auto chạy
* Vì sài đồ free hoặc nếu có trả phí thì cũng sẽ bị giới hạn các repo build cùng lúc nêu

```
Chờ đợi là hạnh phúc. Ahihi
```

* Các trigger phức tạp thì sẽ để phần sau nha. :-)

#### 3.4 Cấu hình lại Project

* Đời không đơn giản khi bạn cấu hình xong `yaml file` là CI sẽ chạy. Lỗi đầu tiên gặp như sau:

```
The command "set -o pipefail && xcodebuild -workspace TheLastProject.xcworkspace -scheme TheLastProject_Development build test | xcpretty" exited with 66
```

* Bạn nên xem lại project đã có các target `test` chưa, nếu chưa có thì:
  * Tạo thêm các target test (Unit & UI)
  * Config build cho tụi nó
  * Check lại 1 lượt các scheme
* Tiếp theo lỗi như thế này

```
error: No profiles for 'com.fx.TheLastProject' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.fx.TheLastProject'. Automatic signing is disabled and unable to generate a profile. To enable automatic signing, pass -allowProvisioningUpdates to xcodebuild. (in target 'TheLastProject')
```

* Chúc mừng là đã thiếu profile `dev, certificate, provisioning profile` … Nôm na là phần `dev signing` để có thể build đc app. Khắc phục sau:
  * Cần phải tắt auto signing
  * Nếu tắt rồi mà không đc thì xoá `team developer`
  * Nếu như vậy mà không đc nữa thì chú ý phần **xcode_destination**, cấu hình lại cho `simulator` chạy. Vì đôi khi một số xcode của anh/chị/em đã ăn với account dev nào rồi nên pó tay thôi.

```
Đừng lo lắng về phần này quá. Vì sẽ được giải quyết tiếp ở phần sau. Ahuhu
```

#### 3.5 Làm đẹp YAML

* **cache** : vì các lệnh `bundler` và `pod` chạy khá tốn thời gian, nên mình sẽ cache lại phần này. Có thể chọn đường dẫn để trỏ tới thư mục cache

```ruby
cache:
  - bundler
  - cocoapods
```

* **script**: dùng để chạy các tập lệnh phức tạp hơn chút
* **install**: thêm các tiện ích khác.

Tạm thời như vậy, sau đây là `yaml file` cơ bản của mình. Ahihi

```ruby
language: objective-c
osx_image: xcode10.1
cache:
  - bundler
  - cocoapods
xcode_workspace: TheLastProject.xcworkspace
xcode_scheme: TheLastProject_Development
xcode_destination: platform=iOS Simulator,OS=12.1,name=iPhone X
before_install:
  - bundle install
  - bundle exec pod install --repo-update
install:
  - set -o pipefail
  - sudo systemsetup -settimezone Asia/Ho_Chi_Minh
  - bundle install --path=vendor/bundle --jobs 4 --retry 3
```

**WELL DONE!**

---



### 4. Auto Review Code and Report

#### 4.1 Sử dụng SwiftLint

* **Khái niệm:**
  * SwiftLint: 
    * Là một **tool** dùng để kiểm tra code (hay gọi là *review code*) theo **conventions** của Swift
    * Có thể bắt gặp nhiều *Lint* khác cho nhiều loại ngôn ngữ khác
    * Về conventions thì mình có thể sử dụng các `rule` mặc định hoặc có thể tự `custom` theo conventions của team hay của dự án
    * Link: <https://github.com/realm/SwiftLint>
* **Cài đặt** dùng `CocoaPod`

```ruby
pod 'SwiftLint'
```

* Thêm script vào Xcode để chạy `swiftlint`

  * Vào `Build Phases` > `New Run Script Phase` 
  * Đoạn script như sau:

  ```ruby
  "${PODS_ROOT}/SwiftLint/swiftlint"
  ```

  Xem hình dưới đây cho rõ

![script_swiftlint_xcode](/images/1.png)

​	Build project và sẽ thấy `Swiftlint` thông báo các `warning` và `error`

#### 4.2 Sử dụng SwiftLint với CI

* Để sử dụng được `SwiftLint` với CI hay ở đây là `Travis CI` thì:
  * Tạo 1 `yaml file` và đặt ở thư mục gốc. CI sẽ tự động quét và chạy
  * Đặt tên sau `.swiftlint.yml`
* Nội dung file tham khảo

```ruby
disabled_rules: # rule identifiers to exclude from running
  - colon
  - comma
  - control_statement
opt_in_rules: # some rules are only opt-in
  - empty_count
  # Find all the available rules by running:
  # swiftlint rules
excluded: # paths to ignore during linting. Takes precedence over `included`.
  - Carthage
  - Pods
  - vendor
analyzer_rules: # Rules run by `swiftlint analyze` (experimental)
  - explicit_self

# configurable rules can be customized from this configuration file
# binary rules can set their severity level
force_cast: warning # implicitly
force_try:
  severity: warning # explicitly
# rules that have both warning and error levels, can set just the warning level
# implicitly
line_length: 110
# they can set both implicitly with an array
type_body_length:
  - 300 # warning
  - 400 # error
# or they can set both explicitly
file_length:
  warning: 500
  error: 1200
# naming rules can set warnings/errors for min_length and max_length
# additionally they can set excluded names
type_name:
  min_length: 4 # only warning
  max_length: # warning and error
    warning: 40
    error: 50
  excluded: iPhone # excluded via string
identifier_name:
  min_length: # only min_length
    error: 3 # only error
  excluded: # excluded via string array
    - id
    - URL
    - GlobalAPIKey
reporter: "xcode" # reporter type (xcode, json, csv, checkstyle, junit, html, emoji, sonarqube, markdown)
```

* Ý nghĩa các từ khoá và lệnh trong file trên thì các bạn từ tìm hiểu và nói chung thì cũng không khó mấy, nên vọc vọc tí là ra
* **Chú ý:**
  * Rule : các luật (rule) cho conventions
  * Thư mục và đường dẫn : nếu bạn có sử dụng thì hãy check, vì ghi thư mục trong file nhưng nó ko có trong repo thì không `pass` đc CI
* Các trường hợp chạy `SwiftLint` oke thì bản build sẽ được kiểm tra, nếu bản build `failed` thì sẽ báo như sau:

```
Testing failed:
	Identifier Name Violation: Variable name should be between 4 and 40 characters long: 'key' (identifier_name)
	Testing cancelled because the build failed.
** TEST FAILED **
```

​	Còn đúng thì âm thầm không show ra gì, nên cứ bình tĩnh mà tiến tới tiếp bước sau, ahihi

#### 4.3 Report/Comment lên GitHub

​	Để có thể post/comment các kết quả của `SwiftLint` lên GitHub thì chúng ta sử dụng `LinterBot`

* **Khái niệm**

  * **Linterbot**
    *  Là 1 con `BOT` , nhiệm vụ của nó là đọc `out put` của `SwiftLint` và comment vào `pull request` trên GitHub
    * Cần có được `access_token` từ account của dev mới có thể post comment vào
  * Link tham khảo: https://github.com/guidomb/linterbot

* **Cài đặt:**

  * Thêm vào `GemFile`

  ```ruby
  gem 'linterbot'
  ```

  * Install

  ```ruby
  gem install linterbot
  ```

* **Sử dụng ở Local**

  ```ruby
  swiftlint lint --reporter json | linterbot REPOSITORY PULL_REQUEST_NUMBER
  ```

* **Sử dụng với Travis CI**

  * Create **bash** script (bash là gì thì tự tìm hiểu nha)

    * Tạo thư mục `bin` (trong thư mục gốc) > tạo 1 file (**ko có đuôi**) và edit như sau:

    ```bash
    #!/bin/bash
    
    if [ "$TRAVIS_PULL_REQUEST" != "false" ]
    then
      ./Pods/SwiftLint/swiftlint --reporter json > swiftlint-report.json || false
      linterbot $TRAVIS_REPO_SLUG $TRAVIS_PULL_REQUEST < swiftlint-report.json
    fi
    ```

    * Chú ý đường dẫn sau `./Pods/SwiftLint/swiftlint` Vì đó là nơi mà mình `pod` swiftlint về nên trỏ vào đúng nó mới chạy đc. Lỗi ngu ngu này mất cả tiếng đồng hồ nếu copy dán không suy nghĩ.
    * Saved file lại với tên là `linter`
    * Edit quyền để tránh các lỗi về `Permission denied` như sau (khi chạy trên CI)

    ```
    0.01s$ ./bin/linter
    /Users/travis/.travis/functions: line 104: ./bin/linter: Permission denied
    The command “./bin/linter” exited with 126.
    ```

    * Lệnh set lại quyền truy cập như sau (dành cho các thanh niên không biết)

    ```bash
     chmod +x ./bin/linter
    ```

  * Edit `.travis.yml`

    * Thêm các lệnh cài đặt `linterbot` 

    ```ruby
    before_install:
      - bundle install
      - gem install linterbot
      - bundle exec pod install --repo-update
    ```

    * Thêm lệnh thực hiện `script` linter

    ```ruby
    script:
      - ./bin/linter
    ```

  * File `.travis.yml` đẹp như sau:

  ```ruby
  language: objective-c
  osx_image: xcode10.1
  cache:
    - bundler
    - cocoapods
  xcode_workspace: TheLastProject.xcworkspace
  xcode_scheme: TheLastProject_Development
  xcode_destination: platform=iOS Simulator,OS=12.1,name=iPhone X
  before_install:
    - bundle install
    - gem install linterbot
    - bundle exec pod install --repo-update
  install:
    - set -o pipefail
    - sudo systemsetup -settimezone Asia/Ho_Chi_Minh
    - bundle install --path=vendor/bundle --jobs 4 --retry 3
  script:
    - ./bin/linter
  ```

* **access_key**

  * Để có thể `post/comment` lên được GitHub thì cần có quyền truy cập từ GitHub
  * Tham khảo cách tạo `access_key` :  <https://help.github.com/en/articles/creating-a-personal-access-token-for-the-command-line>
  * Để đảm bảo tính **bảo mật** thì các `key` sẽ được lưu trữ trên trang `Travis CI`.  Vào setting > tạo 1 `GITHUB_ACCESS_TOKEN`

  ![script_swiftlint_xcode](./images/2.png)

#### 4.4 Kết quả

* Tiến hành commit và gởi pull request để CI chạy và kiểm tra. Nếu đầy đủ các kết quả sau trên `pull request` thì chúc mừng bạn đã pass qua bước này, ahihi
  * Pass đc CI
  * Xuất hiện Comments của `SwiftLint` thông qua `linterbot` 

![script_swiftlint_xcode](./images/3.png)

---

> Thiêu đốt thân tàn ta
>  Lửa thánh cháy bừng bừng,
>  Sống chẳng có gì vui,
>  Thì chết có gì khổ?
>
> Nguyện hành thiện trừ ác.
>  Làm sao cho Quang Minh
>  Bao hỉ lạc bi sầu,
>  Đều hóa thành cát bụi.
>
> Thương thay cho con người,
>  Sao lo buồn lắm vậy?
>  Thương thay cho con người,
>  Sao lo buồn lắm vậy? !
>
> ​	**(Thánh ca Minh Giáo - Ỷ Thiên Đồ Long Ký - Kim Dung)**

---

Chúc mừng bạn đã tới được đây thì đã hoàn thành 1 nữa con đường cho CI (auto build, review code và report lên GitHub). Với các dự án không yêu cầu gì đặc biệt thì như vậy cũng đã ổn rồi. Bạn có thể xông pha ra giang hồ giết quái.

Còn muốn úp level tiếp thì mời bạn tiếp tục phần còn lại là **CD**. Bạn đi pha cốc trà/cafe ăn miếng bánh rồi bình tĩnh làm từng xí 1. Phần này cực kì nỗ não à. Ahihi

---

### 5. Certificate và Provisioning profile

Muốn có được tụi này thì bạn phải bỏ tiền là mua account **Apple Developer Program**, với giá là 99$/năm. Khá chua cho cuộc tình nhưng theo lời khuyên chân thành thì cứ nên đầu tư 1 cái, vì mình còn có nhiều thứ làm với account đó nữa. 

Xin cảm ơn **Viblo** vì đã giải thích quá rõ : <https://viblo.asia/p/khai-quat-ve-apple-developer-account-certificates-provioning-profiles-RQqKLgap57z>

#### 5.1. Certificates

* Sau khi đã join vào Membership Program của Apple, điều bạn cần quan tâm đến là Certificates, App IDs và Provisioning Profiles. Certificate có thể coi là chứng minh nhân dân hoặc bằng lái xe của bạn vậy. 
* Bạn cần request, download và sử dụng các signing certificate để xác thực, chứng nhận bạn có quyền phát triển và phát hành ứng dụng với tài khoản Apple developer tương ứng. 
* Certificate được quản lý dưới dạng 
  * public key 
  * private key. 

* Khi muốn share certificate, bạn cần export certificate (private key) dưới dạng file **.p12**. Nếu làm mất private key, bạn phải tạo lại certificate khác từ đầu. 
* Có 2 loại certificate cần chú ý. Đó là:
  * **Development Certificates**  : Là những certificate dành cho giai đoạn phát triển ứng dụng. Gồm 2 loại chính:
    * **iOS App Development**: Development Certificate được dùng trong quá trình develop app. Giúp developer có thể run và debug app trong môi trường development. Thường thì mỗi thành viên trong team phát triển sẽ cần tạo một certificate loại này để chứng nhận quyền develop app. Mỗi member được tạo tối đa 4 certificate development.
    * **Apple Push Notification service SSL (Sandbox)**: dùng cho server push notification ở môi trường development cho một App ID cụ thể. Ví dụ bạn cần push notification về một app có App ID là com.xxx.yyy thì server push cần được cấp quyền push. Quyền push được thể hiện qua certificate dạng này.
  * Production Certificates : Là những certificate dùng để phát hành ứng dụng. Có nhiều loại, nhưng thường thì ta chỉ cần quan tâm đến 3 loại:
    * **App Store and Ad Hoc**: Certificate loại này cho phép bạn có quyền distribute app lên App Store hoặc trên một list các thiết bị có sẵn trong tài khoản dạng Ad Hoc.
    * **Apple Push Notification service SSL (Production)**: Tương tự như **Apple Push Notification service SSL (Sandbox)**, certificate này cấp quyền cho server push notification đến app trong môi trường production.
    * **In-House and Ad Hoc**: Certificate này chỉ xuất hiện với tài khoản enterprise, cho phép phát hành app dạng Ad Hoc hoặc In-House (không giới hạn số lượng thiết bị, không cần biết trước UUID...).

![cert](https://viblo.asia/uploads/d77e6d52-39d5-4f9f-bd40-ed059e5f406c.png)

#### 5.2. App IDs, Device and Provisioning Profile

* **App IDs** : là định danh các ứng dụng thông mà 1 chuỗi ký tự để đảm bảo là duy nhất mà giang hồ hay gọi là bundle id
* **Devices**: các thiết bị được thêm vào với mã UUID của nó vào account dev. Có như vậy thì bạn mới có thể cài đc app lên thiết bị của bạn với app build ở môi trường distribution.
* **Provisioning Profile:**
  * Một cách ngắn gọn để giải thích về provisioning profile là công thức sau: 
    * **Provisioning Profile = Certificate (Development hoặc Production) + App ID + Devices List (nếu distribute dạng Ad Hoc hoặc debug development).** 
  * Provisioning Profile có thể được tạo ra tự động nếu bạn enable option **Auto Signing** trên Xcode hoặc có thể được tạo và quản lý bởi các tài khoản admin trở lên. 
  * Để tạo provisioning profile cho một app ở môi trường development, bạn cần chọn App ID, chọn các Certificate được join vào, chọn list devices. 
  * Một trong các thành phần của provisioning profile bị invalid thì profile đó sẽ **invalid** và không sử dụng được. 
    * *Ví dụ* một team member *revoke development certificate* của người đó thì tất cả các provisioning profile chứa certificate đó sẽ bị invalid. Hoặc khi xóa một App ID thì tất cả các provisioning profile liên quan cũng sẽ bị xóa theo

![provisioning](https://viblo.asia/uploads/4d6c18ab-3171-4c45-aab9-2a5327eb4e2d.png)

#### 5.3. Thêm các file Certificates và Provisioning vào CI

Bạn cần hiểu nôm na là CI sẽ chạy 1 con máy MAC OS trên server và con máy đó không có gì hết, công việc của mình là sẽ thêm bằng tay tất cả những thứ cần cho việc build vào con CI đó.

* **Chuẩn bị các file:**

  * Tạo 1 app id với bundle id của project của bạn
  * 2 file certificates cho 2 môi trường Development và Distribution
    * Public key là các file .cer
    * Private key là các file .p12 (tụi nó có pass)
  * 2 file provisioning cho 2 môi trường đó

  Các bạn có thể sử dụng các file của mình trong repo này, tuy nhiên sau này nếu không chạy đc thì do mình đã xoá trên account dev của mình. Vì vậy tốt nhất là nên sài đồ của bản thân cho chắc.


* **Cấu hình lại project** của các bạn để có thể build và test đc project bằng Xcode với 2 kiểu sau:

  * Auto signing
  * Dùng Provisioning và Certificate

* **Coding**

  * Di chuyển script `linter` vào thư mục `./script` để tiện quản lý
  * Viết scripts để thêm các file trên vào máy ảo Travis CI. Bạn tạo 1 shell bash tên là `add-key.sh` trong thư mục `./scripts`

  ```bash
  #!/bin/bash
  
  KEY_CHAIN=ios-build.keychain
  security create-keychain -p travis $KEY_CHAIN
  # Make the keychain the default so identities are found
  security default-keychain -s $KEY_CHAIN
  # Unlock the keychain
  security unlock-keychain -p travis $KEY_CHAIN
  # Set keychain locking timeout to 3600 seconds
  security set-keychain-settings -t 3600 -u $KEY_CHAIN
  
  # Add certificates to keychain and allow codesign to access them
  security import ./scripts/certs/dis.cer -k $KEY_CHAIN -T /usr/bin/codesign
  security import ./scripts/certs/dev.cer -k $KEY_CHAIN -T /usr/bin/codesign
  
  security import ./scripts/certs/dis.p12 -k $KEY_CHAIN -P 12345678  -T /usr/bin/codesign
  security import ./scripts/certs/dev.p12 -k $KEY_CHAIN -P 12345678  -T /usr/bin/codesign
  
  security set-key-partition-list -S apple-tool:,apple: -s -k travis ~/Library/Keychains/ios-build.keychain
  
  echo "list keychains: "
  security list-keychains
  echo " ****** "
  
  echo "find indentities keychains: "
  security find-identity -p codesigning  ~/Library/Keychains/ios-build.keychain
  echo " ****** "
  
  # Put the provisioning profile in place
  mkdir -p ~/Library/MobileDevice/Provisioning\ Profiles
  
  uuid=`grep UUID -A1 -a ./scripts/profiles/TheLastProject_Dev.mobileprovision | grep -io "[-A-F0-9]\{36\}"`
  cp ./scripts/profiles/TheLastProject_Dev.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/$uuid.mobileprovision
  
  uuid=`grep UUID -A1 -a ./scripts/profiles/TheLastProject_AdHoc.mobileprovision | grep -io "[-A-F0-9]\{36\}"`
  cp ./scripts/profiles/TheLastProject_AdHoc.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/$uuid.mobileprovision
  
  #cp ./scripts/profiles/TheLastProject_Dev.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/
  #cp ./scripts/profiles/TheLastProject_AdHoc.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/
  
  echo "show provisioning profile"
  ls ~/Library/MobileDevice/Provisioning\ Profiles
  echo " ****** "
  ```

  * Edit lại file `.travis.yml` để chạy các shell bass này

  ```bash
  before_script:
    - chmod a+x ./scripts/add-key.sh
    - sh ./scripts/add-key.sh
  script:
    - ./scripts/linter
  ```



#### 5.4. Giải thích

* Tạo `keychain`
  * Là nơi để chứa các certificate và provisioning trong MAC OS
  * MAC OS có thể tìm nhanh chóng và bạn có thể dùng `ctrl + space` sẽ thấy
  * Tạo 1 keychain với:
    * Tên: `ios-build.keychain`
    * Pass: `travis`
  * Phần này cho các thanh niên chuyên copy dán mà lỡ chạy bậy thì còn có `pass` để mà xoá

```bash
KEY_CHAIN=ios-build.keychain
security create-keychain -p travis $KEY_CHAIN
```

```bash
# Make the keychain the default so identities are found
security default-keychain -s $KEY_CHAIN
# Unlock the keychain
security unlock-keychain -p travis $KEY_CHAIN
# Set keychain locking timeout to 3600 seconds
security set-keychain-settings -t 3600 -u $KEY_CHAIN
```

* Set thuộc tính có keychain
  * Default : để MAC OS tìm trước
  * Unlock chúng để khỏi phải nhập pass nếu truy xuất tới
  * Time out cho nó `1 giờ` để tự huỹ, chứ cài nhiều quá loạn
* Thêm các file **certificates**  vào
  * Add lần lượt từng cái vào từng thư mục
  * 2 em file cert ko cần add cũng đc, chủ yếu là 2 em .p12
  * Pass của file p12 là `12345678`

```bash
# Add certificates to keychain and allow codesign to access them
security import ./scripts/certs/dis.cer -k $KEY_CHAIN -T /usr/bin/codesign
security import ./scripts/certs/dev.cer -k $KEY_CHAIN -T /usr/bin/codesign

security import ./scripts/certs/dis.p12 -k $KEY_CHAIN -P 12345678  -T /usr/bin/codesign
security import ./scripts/certs/dev.p12 -k $KEY_CHAIN -P 12345678  -T /usr/bin/codesign
```

* Tuy nhiên đời không phải là mơ, nếu bạn chạy với Travis CI là MAC OS, thì đôi khi add đc rồi nhưng lại không hoạt động.

fastlane/release_notes.txt
Appstore: 50fc6d5 typing


scripts/add-key.sh
#!/bin/bash

KEY_CHAIN=ios-build.keychain
security create-keychain -p travis $KEY_CHAIN
# Make the keychain the default so identities are found
security default-keychain -s $KEY_CHAIN
# Unlock the keychain
security unlock-keychain -p travis $KEY_CHAIN
# Set keychain locking timeout to 3600 seconds
security set-keychain-settings -t 3600 -u $KEY_CHAIN

# Add certificates to keychain and allow codesign to access them
#security import ./scripts/certs/dis.cer -k $KEY_CHAIN -T /usr/bin/codesign
#security import ./scripts/certs/dev.cer -k $KEY_CHAIN -T /usr/bin/codesign

security import ./scripts/certs/dis.p12 -k $KEY_CHAIN -P 12345678  -T /usr/bin/codesign
security import ./scripts/certs/dev.p12 -k $KEY_CHAIN -P 12345678  -T /usr/bin/codesign

security set-key-partition-list -S apple-tool:,apple: -s -k travis ~/Library/Keychains/ios-build.keychain

echo "list keychains: "
security list-keychains
echo " ****** "

echo "find indentities keychains: "
security find-identity -p codesigning  ~/Library/Keychains/ios-build.keychain
echo " ****** "

# Put the provisioning profile in place
mkdir -p ~/Library/MobileDevice/Provisioning\ Profiles

uuid=`grep UUID -A1 -a ./scripts/profiles/TheLastProject_Dev.mobileprovision | grep -io "[-A-F0-9]\{36\}"`
cp ./scripts/profiles/TheLastProject_Dev.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/$uuid.mobileprovision

uuid=`grep UUID -A1 -a ./scripts/profiles/TheLastProject_AdHoc.mobileprovision | grep -io "[-A-F0-9]\{36\}"`
cp ./scripts/profiles/TheLastProject_AdHoc.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/$uuid.mobileprovision

#cp ./scripts/profiles/TheLastProject_Dev.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/
#cp ./scripts/profiles/TheLastProject_AdHoc.mobileprovision ~/Library/MobileDevice/Provisioning\ Profiles/

echo "show provisioning profile"
ls ~/Library/MobileDevice/Provisioning\ Profiles
echo " ****** "